### ISSUE TITLE
TypeError: 'NoneType' object is not iterable with module apt autoremove

### ISSUE THREAD

**Issue by scjody** [ansible/ansible#21472]
_From @galindro on June 23, 2016 23:4_

##### ISSUE TYPE
- Bug Report
##### COMPONENT NAME

apt autoremove
##### ANSIBLE VERSION

``` bash
ansible 2.1.0.0
  config file = /home/galindro/git_repos/devops/ansible.cfg
  configured module search path = Default w/o overrides
```
##### CONFIGURATION

``` bash
$ cat ansible.cfg

[defaults]
host_key_checking = False

[ssh_connection]
pipelining=False
ssh_args = -o ForwardAgent=yes -o ControlMaster=auto -o ControlPersist=30m -o ControlPath="~/.ansible/tmp/ansible-ssh-%h-%p-%r"
```

I'm NOT using any ANSIBLE_\* environment variable
##### OS / ENVIRONMENT

```
No LSB modules are available.
Distributor ID: Debian
Description:    Debian GNU/Linux 8.5 (jessie)
Release:    8.5
Codename:   jessie
```
##### SUMMARY

I'm used the apt module with autoremove option and this error was showed when I run the task:

```
TASK [common : autoremove unused packages] *************************************
fatal: [172.30.2.214]: FAILED! => {"changed": false, "failed": true, "module_stderr": "", "module_stdout": "Traceback (most recent call last):\r\n  File \"/tmp/ansible_RpjDPt/ansible_module_apt.py\", line 842, in <module>\r\n    main()\r\n  File \"/tmp/ansible_RpjDPt/ansible_module_apt.py\", line 802, in main\r\n    for package in packages:\r\nTypeError: 'NoneType' object is not iterable\r\n", "msg": "MODULE FAILURE", "parsed": false}
```
##### STEPS TO REPRODUCE

Create this task:

``` yaml
- name: autoremove unused packages
  apt:
    autoremove: yes
  sudo: yes
  tags:
    - infra
```

Put that task inside a simple role

Create a playbook:

``` yaml
- hosts: all
  role:
     - myrolename
  environment:
    TMOUT: 600
```

Run the playbook

```
ansible-playbook -i ec2.py -v -b myplay.yml
```

P.S.: my host is an instance running on amazon ec2
##### EXPECTED RESULTS

This error:

``` bash
TASK [myrolename : autoremove unused packages] *************************************
fatal: [172.30.2.214]: FAILED! => {"changed": false, "failed": true, "module_stderr": "", "module_stdout": "Traceback (most recent call last):\r\n  File \"/tmp/ansible_RpjDPt/ansible_module_apt.py\", line 842, in <module>\r\n    main()\r\n  File \"/tmp/ansible_RpjDPt/ansible_module_apt.py\", line 802, in main\r\n    for package in packages:\r\nTypeError: 'NoneType' object is not iterable\r\n", "msg": "MODULE FAILURE", "parsed": false}
```
##### ACTUAL RESULTS

Running with -vvvv:

```
TASK [myrolename : autoremove unused packages] *************************************
task path: /home/galindro/git_repos/devops/roles/myrolename/tasks/debian.yml:52
<172.30.2.214> ESTABLISH SSH CONNECTION FOR USER: ubuntu
<172.30.2.214> SSH: EXEC ssh -C -vvv -o ForwardAgent=yes -o ControlMaster=auto -o ControlPersist=30m -o 'ControlPath=~/.ansible/tmp/ansible-ssh-%h-%p-%r' -o StrictHostKeyChecking=no -o Port=40 -o 'IdentityFile="sbkey_20150204_app.pem"' -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=ubuntu -o ConnectTimeout=10 172.30.2.214 '/bin/sh -c '"'"'( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1466722720.22-229804017504254 `" && echo ansible-tmp-1466722720.22-229804017504254="` echo $HOME/.ansible/tmp/ansible-tmp-1466722720.22-229804017504254 `" ) && sleep 0'"'"''
<172.30.2.214> PUT /tmp/tmpBiapsW TO /home/ubuntu/.ansible/tmp/ansible-tmp-1466722720.22-229804017504254/apt
<172.30.2.214> SSH: EXEC sftp -b - -C -vvv -o ForwardAgent=yes -o ControlMaster=auto -o ControlPersist=30m -o 'ControlPath=~/.ansible/tmp/ansible-ssh-%h-%p-%r' -o StrictHostKeyChecking=no -o Port=40 -o 'IdentityFile="sbkey_20150204_app.pem"' -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=ubuntu -o ConnectTimeout=10 '[172.30.2.214]'
<172.30.2.214> ESTABLISH SSH CONNECTION FOR USER: ubuntu
<172.30.2.214> SSH: EXEC ssh -C -vvv -o ForwardAgent=yes -o ControlMaster=auto -o ControlPersist=30m -o 'ControlPath=~/.ansible/tmp/ansible-ssh-%h-%p-%r' -o StrictHostKeyChecking=no -o Port=40 -o 'IdentityFile="sbkey_20150204_app.pem"' -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=ubuntu -o ConnectTimeout=10 -tt 172.30.2.214 '/bin/sh -c '"'"'sudo -H -S -n -u root /bin/sh -c '"'"'"'"'"'"'"'"'echo BECOME-SUCCESS-yhgovfdopoujnzwegznjpbbaoqcpcnyd; LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 TMOUT=600 LC_MESSAGES=en_US.UTF-8 /usr/bin/python /home/ubuntu/.ansible/tmp/ansible-tmp-1466722720.22-229804017504254/apt; rm -rf "/home/ubuntu/.ansible/tmp/ansible-tmp-1466722720.22-229804017504254/" > /dev/null 2>&1'"'"'"'"'"'"'"'"' && sleep 0'"'"''
fatal: [172.30.2.214]: FAILED! => {"changed": false, "failed": true, "invocation": {"module_name": "apt"}, "module_stderr": "OpenSSH_6.7p1 Debian-5+deb8u2, OpenSSL 1.0.1t  3 May 2016\r\ndebug1: Reading configuration data /home/galindro/.ssh/config\r\ndebug1: Reading configuration data /etc/ssh/ssh_config\r\ndebug1: /etc/ssh/ssh_config line 19: Applying options for *\r\ndebug1: auto-mux: Trying existing master\r\ndebug2: fd 3 setting O_NONBLOCK\r\ndebug2: mux_client_hello_exchange: master version 4\r\ndebug3: mux_client_forwards: request forwardings: 0 local, 0 remote\r\ndebug3: mux_client_request_session: entering\r\ndebug3: mux_client_request_alive: entering\r\ndebug3: mux_client_request_alive: done pid = 32546\r\ndebug3: mux_client_request_session: session request sent\r\ndebug1: mux_client_request_session: master session id: 2\r\ndebug3: mux_client_read_packet: read header failed: Broken pipe\r\ndebug2: Received exit status from master 0\r\nShared connection to 172.30.2.214 closed.\r\n", "module_stdout": "Traceback (most recent call last):\r\n  File \"/tmp/ansible_uvVsjU/ansible_module_apt.py\", line 842, in <module>\r\n    main()\r\n  File \"/tmp/ansible_uvVsjU/ansible_module_apt.py\", line 802, in main\r\n    for package in packages:\r\nTypeError: 'NoneType' object is not iterable\r\n", "msg": "MODULE FAILURE", "parsed": false}
```

_Copied from original issue: ansible/ansible-modules-core#4029_

**Comment by scjody** [comment]
_From @mrgnr on June 26, 2016 0:33_

I'm seeing the same error as @galindro when running this task:

```
- name: Remove unused dependencies
  apt: autoremove=yes
```

However, if I specify a value for the `name` parameter, the task succeeds:

```
- name: Remove unused dependencies
  apt: autoremove=yes name=htop
```

The [apt module documentation](https://docs.ansible.com/ansible/apt_module.html) says that `name` is _not_ a required parameter, and it makes sense that we might want to autoremove packages without installing anything.

**Comment by scjody** [comment]
_From @domdom82 on July 5, 2016 14:21_

Ran into the same issue today. +1 to fix this.

Using this as a workaround:

```
shell: apt-get -y autoremove --purge 
```

**Comment by scjody** [comment]
_From @peterkuiper on July 14, 2016 14:59_

I'm having the exact same problem. My ansible version:

ansible 2.1.0.0

This task also doesn't work:

```
- name: Update apt packages
  apt: upgrade=dist
  autoremove=yes
  update_cache=yes
  cache_valid_time=3600
```

The solution proposed by @mrgnr works.

**Comment by scjody** [comment]
_From @ansibot on July 30, 2016 14:18_

@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/jctanner/ansibullbot/blob/ISSUE_TRIAGE_CLASS/ISSUE_HELP.md)

**Comment by scjody** [comment]
_From @basictheprogram on August 29, 2016 20:4_

Issue is in devel too.

$ ansible --version
ansible 2.2.0 (devel 8210260d94)

**Comment by scjody** [comment]
_From @ansibot on August 29, 2016 20:7_

@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by scjody** [comment]
_From @Blake- on September 16, 2016 17:44_

output of -vvvv
ansible 2.1.1.0

```
<ip> ESTABLISH SSH CONNECTION FOR USER: user
<ip> SSH: EXEC ssh -C -vvv -o ControlMaster=auto -o ControlPersist=60s -o Port=922 -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=user -o ConnectTimeout=10 -o ControlPath=/home/ansible/.ansible/cp/ansible-ssh-%h-%p-%r ip '/bin/sh -c '"'"'sudo -H -S  -p "[sudo via ansible, key=key] password: " -u root /bin/sh -c '"'"'"'"'"'"'"'"'echo BECOME-SUCCESS; LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 LC_MESSAGES=en_US.UTF-8 /usr/bin/python'"'"'"'"'"'"'"'"' && sleep 0'"'"''
fatal: [ip]: FAILED! => {"changed": false, "failed": true, "invocation": {"module_name": "apt"}, "module_stderr": "OpenSSH_6.6.1, OpenSSL 1.0.1f 6 Jan 2014\r\ndebug1: Reading configuration data /etc/ssh/ssh_config\r\ndebug1: /etc/ssh/ssh_config line 19: Applying options for *\r\ndebug1: auto-mux: Trying existing master\r\ndebug2: fd 3 setting O_NONBLOCK\r\ndebug2: mux_client_hello_exchange: master version 4\r\ndebug3: mux_client_forwards: request forwardings: 0 local, 0 remote\r\ndebug3: mux_client_request_session: entering\r\ndebug3: mux_client_request_alive: entering\r\ndebug3: mux_client_request_alive: done pid = 18506\r\ndebug3: mux_client_request_session: session request sent\r\ndebug1: mux_client_request_session: master session id: 2\r\nTraceback (most recent call last):\n  File \"/tmp/ansible_s8Bnka/ansible_module_apt.py\", line 843, in <module>\n    main()\n  File \"/tmp/ansible_s8Bnka/ansible_module_apt.py\", line 803, in main\n    for package in packages:\nTypeError: 'NoneType' object is not iterable\ndebug3: mux_client_read_packet: read header failed: Broken pipe\r\ndebug2: Received exit status from master 1\r\n", "module_stdout": "", "msg": "MODULE FAILURE", "parsed": false}
```

**Comment by scjody** [comment]
_From @ansibot on September 16, 2016 21:32_

@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by scjody** [comment]
_From @chrpinedo on September 26, 2016 9:9_

I also confirm the bug with ansible 2.1.1.0

**Comment by scjody** [comment]
_From @ansibot on October 6, 2016 13:59_

@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by scjody** [comment]
_From @shay-li on October 9, 2016 11:59_

+1 to fix this issue

**Comment by scjody** [comment]
_From @hlpinform on October 21, 2016 8:7_

+1 to fix this issue, we ran into this issue today with this version:

ansible 2.3.0 (devel d9b570aaae) last updated 2016/10/17 09:47:20 (GMT +000)
  lib/ansible/modules/core: (detached HEAD 15f8c6835b) last updated 2016/10/17 09:47:24 (GMT +000)
  lib/ansible/modules/extras: (detached HEAD 4ca8e5347a) last updated 2016/10/17 09:47:28 (GMT +000)

**Comment by scjody** [comment]
_From @ansibot on October 22, 2016 8:50_

@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by scjody** [comment]
_From @etienne-me on October 25, 2016 14:28_

+1

**Comment by scjody** [comment]
_From @gdelpierre on November 3, 2016 8:52_

Same as @mrgnr , name is not a required parameter.

**Comment by scjody** [comment]
_From @ansibot on November 6, 2016 9:46_

@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by scjody** [comment]
Note that https://github.com/ansible/ansible-modules-core/issues/4029#issuecomment-230492142 is not a good workaround. If you end up with a package that defaults to interactive configuration on removal (I've seen this with kernel images), the task will hang.

https://github.com/ansible/ansible-modules-core/issues/4029#issuecomment-228577588 works for me, except I use `bash` because we don't install htop.

**Comment by scjody** [comment]
_From @tdely on November 16, 2016 13:19_

https://github.com/ansible/ansible-modules-core/issues/4029#issuecomment-228577588 works for me as well, but only when the package specified is installed or removed; if there is no change to the package no autoremove action is taken.

**Comment by scjody** [comment]
_From @ansibot on November 22, 2016 9:54_

@ansible, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by scjody** [comment]
_From @ansibot on December 9, 2016 19:54_

This repository has been locked. All new issues and pull requests should be filed in https://github.com/ansible/ansible

Please read through the <a href="http://docs.ansible.com/ansible/dev_guide/repomerge.html#move-issues-and-prs-to-new-repo">repomerge</a> page in the dev guide. The guide contains links to tools which automatically move your issue or pull request to the ansible/ansible repo.

**Comment by scjody** [comment]
_From @zahaim on February 14, 2017 6:14_

+1

**Comment by itnick** [comment]
+1
have to use the shell workaround and neglect the warning of "[WARNING]: Consider using apt module rather than running apt-get"

**Comment by gaddman** [comment]
Tested with commit 7966beb6c9529fe86ea9390c27a46d49f3bd968b and it fixes this error.

**Comment by bcoca** [comment]
fixed via  7966beb

**Comment by SethosII** [comment]
I tested the `autoremove` option with `2.3.0.0-1ppa~xenial` from the ansible PPA but it doesn't find any autoremoveable packages (just says OK), allthough there are some old kernel images that can be autoremoved. Should I open a new issue or does it belong to this one?

**Comment by dentarg** [comment]
@bcoca @abadger will all versions (2.0, 2.1, 2.2, 2.3) get the fix, or only 2.3?

**Comment by gyurco** [comment]
@SethosII
I think, yes, it is still buggy. The autoremove=yes results in the command 'apt-get -y --auto-remove install' which doesn't do anything.

**Comment by abadger** [comment]
Unfortunately, reviewing closed issues isn't part of our workflow so we don't see questions or new information that are posted to them.  Opening a new issue is the best way to get attention.  If you think it's a continuation of the old bug, linking to the old bug report will help us to review the history.

2.0, 2.1, and 2.2 are only getting security fixes at this point so the fix would be available in 2.4 (current devel) and possibly 2.3 if we backport it to the latest stable release.

**Comment by SethosII** [comment]
@abadger Thanks, I created a new issue: https://github.com/ansible/ansible/issues/26181.
